src/Swarm/Game/ScenarioInfo.hs

```diff
@@ -0,0 +1,328 @@
+import Swarm.Game.Entity
+import Swarm.Game.Scenario
+import Swarm.Util (getSwarmSavePath)
+import System.Directory (canonicalizePath, doesDirectoryExist, doesFileExist, listDirectory)
+import System.FilePath (pathSeparator, splitDirectories, takeBaseName, takeExtensions, (-<.>), (</>))
+import Witch (into)
+
+-- Some orphan ZonedTime instances
+
+instance Eq ZonedTime where
+  (==) = (==) `on` zonedTimeToUTC
+
+instance Ord ZonedTime where
+  (<=) = (<=) `on` zonedTimeToUTC
+
+-- | A @ScenarioStatus@ stores the status of a scenario along with
+--   appropriate metadata: not started, in progress, or complete.
+--   Note that "in progress" is currently a bit of a misnomer since
+--   games cannot be saved; at the moment it really means more like
+--   "you played this scenario before but didn't win".
+data ScenarioStatus
+  = NotStarted
+  | InProgress
+      { -- | Time when the scenario was started including time zone.
+        _scenarioStarted :: ZonedTime
+      , -- | Time elapsed until quitting the scenario.
+        _scenarioElapsed :: NominalDiffTime
+      , -- | Ticks elapsed until quitting the scenario.
+        _scenarioElapsedTicks :: Integer
+      }
+  | Complete
+      { -- | Time when the scenario was started including time zone.
+        _scenarioStarted :: ZonedTime
+      , -- | Time elapsed until quitting the scenario.
+        _scenarioElapsed :: NominalDiffTime
+      , -- | Ticks elapsed until quitting the scenario.
+        _scenarioElapsedTicks :: Integer
+      }
+  deriving (Eq, Ord, Show, Read, Generic)
+
+instance FromJSON ScenarioStatus where
+  parseJSON = genericParseJSON scenarioOptions
+
+instance ToJSON ScenarioStatus where
+  toEncoding = genericToEncoding scenarioOptions
+  toJSON = genericToJSON scenarioOptions
+
+-- | A @ScenarioInfo@ record stores metadata about a scenario: its
+--   canonical path, most recent status, and best-ever status.
+data ScenarioInfo = ScenarioInfo
+  { _scenarioPath :: FilePath
+  , _scenarioStatus :: ScenarioStatus
+  , _scenarioBest :: ScenarioStatus
+  }
+  deriving (Eq, Ord, Show, Read, Generic)
+
+instance FromJSON ScenarioInfo where
+  parseJSON = genericParseJSON scenarioOptions
+
+instance ToJSON ScenarioInfo where
+  toEncoding = genericToEncoding scenarioOptions
+  toJSON = genericToJSON scenarioOptions
+
+scenarioOptions :: Options
+scenarioOptions =
+  defaultOptions
+    { fieldLabelModifier = map toLower . drop (length "_scenario")
+    }
+
+makeLensesWith (lensRules & generateSignatures .~ False) ''ScenarioInfo
+
+-- | The path of the scenario, relative to @data/scenarios@.
+scenarioPath :: Lens' ScenarioInfo FilePath
+
+-- | The status of the scenario.
+scenarioStatus :: Lens' ScenarioInfo ScenarioStatus
+
+-- | The best status of the scenario.
+scenarioBest :: Lens' ScenarioInfo ScenarioStatus
+
+-- | Update the current @ScenarioInfo@ record when quitting a game.
+updateScenarioInfoOnQuit :: ZonedTime -> Integer -> Bool -> ScenarioInfo -> ScenarioInfo
```

Comment:
Maybe the docstring should mention that the "best" time is either shortest (once completed) or longest (while playing, which works well for Classic mode).

There is a comment in `updateScenarioInfoOnQuit` saying so, but we should agree on the desired behaviour.

Reply:
I added a small explanation:

> Note that when comparing "best" times, shorter is not always better!
As long as the scenario is not completed (e.g. some do not have a win condition) we consider having fun _longer_ to be better.